x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/FrozenCacheService.java

```diff
@@ -169,96 +194,120 @@ public int getRecoveryRangeSize() {
         return toIntBytes(recoveryRangeSize.getBytes());
     }
 
-    private int getRegion(long position) {
-        return Math.toIntExact(position / regionSize);
+    // Number of large regions used when caching a file of the given length
+    private int largeRegions(long fileLength) {
+        final int largeRegions = Math.toIntExact(fileLength / regionSize);
+        final long remainder = fileLength % regionSize;
+        if (remainder == 0) {
+            return largeRegions;
+        }
+        // if we fill up the next region more than 50%, add another region
+        if (remainder > regionSize / 2) {
+            return largeRegions + 1;
+        }
+        final int smallRegionsNeeded = Math.toIntExact(remainder / smallRegionSize);
+        // arbitrary heuristic: don't create more than twice the value of (large regions + 1) to strike a balance between not wasting too
```

Comment:
Not sure about this, this worked out somewhat nicely for some test examples but without real wold data on the eviction patterns I have a hard time guessing an approach here.